src/main/java/solrocr/OcrHighlighter.java

```diff
@@ -394,92 +416,105 @@ public OcrHighlightResult[] highlightOcrFields(
-                    content.getPointer(),
-                    e.getMessage());
+
+          // Final aliases for lambda
+          final int docIdFinal = docId;
+          final int fieldIdxFinal = fieldIdx;
+          final IterableCharSequence contentFinal = content;
+          Runnable hlFn =
+              () -> {
+                try {
+                  highlightDocField(
+                      docIdFinal,
+                      docInIndex,
+                      fieldIdxFinal,
+                      contentFinal,
+                      fieldHighlighter,
+                      leafReader,
+                      snippetLimit,
+                      resultByDocIn,
+                      snippetCountsByField);
+                } catch (ExitingIterCharSeq.ExitingIterCharSeqException
+                    | ExitableDirectoryReader.ExitingReaderException e) {
+                  throw e;
+                } catch (IOException | RuntimeException e) {
+                  // This catch-all prevents OCR highlighting from failing the complete query,
+                  // instead users get an error message in their Solr log.
+                  if (contentFinal.getPointer() != null) {
+                    log.error(
+                        "Could not highlight OCR content for document {} at '{}'",
+                        docIdFinal,
+                        contentFinal.getPointer(),
+                        e);
+                  } else {
+                    log.error(
+                        "Could not highlight OCR for document {} with OCR markup '{}...'",
+                        docIdFinal,
+                        contentFinal.subSequence(0, 256),
+                        e);
+                  }
+                } finally {
+                  if (contentFinal instanceof AutoCloseable) {
+                    try {
+                      ((AutoCloseable) contentFinal).close();
+                    } catch (Exception e) {
+                      log.warn(
+                          "Encountered error while closing content iterator for {}: {}",
+                          contentFinal.getPointer(),
+                          e.getMessage());
+                    }
+                  }
+                }
+              };
+          if (hlThreadPool.getActiveCount() == hlThreadPool.getMaximumPoolSize()) {
+            // Run the task synchronously on the current thread if the pool is fully busy
+            try {
+              hlFn.run();
+            } catch (ExitingIterCharSeq.ExitingIterCharSeqException
+                | ExitableDirectoryReader.ExitingReaderException e) {
+              if (respHeader.get(PARTIAL_OCR_HIGHLIGHTS) == null) {
+                respHeader.put(PARTIAL_OCR_HIGHLIGHTS, Boolean.TRUE);
+                log.warn("OCR Highlighting timed out while handling " + content.getPointer(), e);
               }
+              resultByDocIn[docInIndex] = null;
```

Comment:
Should we break the docLoop at this point? The exceptions mean that a timeout has occurred, correct?

Reply:
Yes, you're right! Makes no sense to queue up further tasks even if the pool should free up a bit in the next iteration, since we've hit the timeout already 👍🏾